File: app/src/main/java/by/offvanhooijdonk/tofreedom/ui/StartActivity.java
package by.offvanhooijdonk.tofreedom.ui;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.TextView;
import android.widget.TimePicker;

import java.util.Calendar;

import by.offvanhooijdonk.tofreedom.R;
import by.offvanhooijdonk.tofreedom.helper.DateFormatHelper;
import by.offvanhooijdonk.tofreedom.helper.PrefHelper;
import by.offvanhooijdonk.tofreedom.ui.countdown.CountdownActivity;

public class StartActivity extends AppCompatActivity implements DatePickerDialog.OnDateSetListener, TimePickerDialog.OnTimeSetListener {

    private TextView txtSetFreedomDate;
    private TextView txtSetFreedomTime;
    private TextView txtPickedTime;
    private Button btnContinue;

    private long freedomTime = PrefHelper.FREEDOM_TIME_DEFAULT;
    private boolean isDateSet = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        long freedomTime = PrefHelper.getFreedomTime(this);
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT) {
            navigateCountdownView();
        } else {
            init();
        }
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
    }

    @Override
    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
        Calendar timeNewValue = Calendar.getInstance();
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT) {
            timeNewValue.setTimeInMillis(freedomTime);
        } else {
            timeNewValue.set(Calendar.HOUR_OF_DAY, 12);
            timeNewValue.set(Calendar.MINUTE, 0);
            timeNewValue.set(Calendar.SECOND, 0);
            timeNewValue.set(Calendar.MILLISECOND, 0);
        }

        timeNewValue.set(Calendar.YEAR, year);
        timeNewValue.set(Calendar.MONTH, month);
        timeNewValue.set(Calendar.DAY_OF_MONTH, dayOfMonth);

        freedomTime = timeNewValue.getTimeInMillis();
        updateFreedomDateView();

        isDateSet = true;
    }

    @Override
    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
        Calendar timeNewValue = Calendar.getInstance();
        timeNewValue.setTimeInMillis(freedomTime);

        timeNewValue.set(Calendar.HOUR_OF_DAY, hourOfDay);
        timeNewValue.set(Calendar.MINUTE, minute);

        freedomTime = timeNewValue.getTimeInMillis();
        updateFreedomDateView();
    }

    private void init() {
        setContentView(R.layout.activity_start);

        txtSetFreedomDate = (TextView) findViewById(R.id.txtSetFreedomDate);
        txtSetFreedomTime = (TextView) findViewById(R.id.txtSetFreedomTime);
        txtPickedTime = (TextView) findViewById(R.id.txtPickedTime);
        btnContinue = (Button) findViewById(R.id.btnContinue);

        txtSetFreedomDate.setOnClickListener(v -> onFreedomDateClick());
        txtSetFreedomTime.setOnClickListener(v -> onFreedomDateTime());
        btnContinue.setOnClickListener(v -> {
            if (isDateSet && freedomTime > System.currentTimeMillis()) {
                PrefHelper.setFreedomTime(StartActivity.this, freedomTime);
                PrefHelper.setCountdownStartDate(StartActivity.this, System.currentTimeMillis());
                navigateCountdownView();
            } else {
                new AlertDialog.Builder(this)
                        .setCancelable(true)
                        .setTitle(R.string.date_warning_title)
                        .setMessage(R.string.date_warning_msg)
                        .setPositiveButton(android.R.string.ok, (dialog, which) -> dialog.dismiss())
                        .show();
            }
        });

        txtPickedTime.setVisibility(View.GONE);
    }

    private void onFreedomDateTime() {
        Calendar now = Calendar.getInstance();
        new TimePickerDialog(this, this,
                now.get(Calendar.HOUR_OF_DAY),
                now.get(Calendar.MINUTE), true).show();
    }

    private void onFreedomDateClick() {
        Calendar now = Calendar.getInstance();
        DatePickerDialog dialog = new DatePickerDialog(this, this,
                now.get(Calendar.YEAR),
                now.get(Calendar.MONTH),
                now.get(Calendar.DAY_OF_MONTH));
        dialog.getDatePicker().setMinDate(now.getTimeInMillis());

        now.add(Calendar.YEAR, PrefHelper.MAX_YEARS_AHEAD);
        dialog.getDatePicker().setMaxDate(now.getTimeInMillis());

        dialog.show();
    }

    private void updateFreedomDateView() {
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT) {
            txtPickedTime.setText(DateFormatHelper.formatForStart(freedomTime));
            txtPickedTime.setVisibility(View.VISIBLE);
        }
    }

    private void navigateCountdownView() {
        startActivity(new Intent(this, CountdownActivity.class));
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
    }
}

File: app/src/main/java/by/offvanhooijdonk/tofreedom/ui/StartActivity.kt
package by.offvanhooijdonk.tofreedom.ui

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.text.format.DateFormat
import android.view.View
import android.widget.DatePicker
import android.widget.TimePicker
import by.offvanhooijdonk.tofreedom.R
import by.offvanhooijdonk.tofreedom.helper.AlarmHelper
import by.offvanhooijdonk.tofreedom.helper.DateFormatHelper
import by.offvanhooijdonk.tofreedom.helper.PrefHelper
import by.offvanhooijdonk.tofreedom.ui.countdown.CountdownActivity
import kotlinx.android.synthetic.main.activity_start.*
import java.util.*

class StartActivity : AppCompatActivity(), DatePickerDialog.OnDateSetListener, TimePickerDialog.OnTimeSetListener {

    private var freedomTime = PrefHelper.FREEDOM_TIME_DEFAULT.toLong()
    private var isDateSet = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val freedomTime = PrefHelper.getFreedomTime(this)
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT.toLong()) {
            navigateCountdownView()
        } else {
            init()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
    }

    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        val timeNewValue = Calendar.getInstance()
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT.toLong()) {
            timeNewValue.timeInMillis = freedomTime
        } else {
            timeNewValue.set(Calendar.HOUR_OF_DAY, 12)
            timeNewValue.set(Calendar.MINUTE, 0)
            timeNewValue.set(Calendar.SECOND, 0)
            timeNewValue.set(Calendar.MILLISECOND, 0)
        }

        timeNewValue.set(Calendar.YEAR, year)
        timeNewValue.set(Calendar.MONTH, month)
        timeNewValue.set(Calendar.DAY_OF_MONTH, dayOfMonth)

        freedomTime = timeNewValue.timeInMillis
        updateFreedomDateView()

        isDateSet = true
    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        val timeNewValue = Calendar.getInstance()
        timeNewValue.timeInMillis = freedomTime

        timeNewValue.set(Calendar.HOUR_OF_DAY, hourOfDay)
        timeNewValue.set(Calendar.MINUTE, minute)

        freedomTime = timeNewValue.timeInMillis
        updateFreedomDateView()
    }

    private fun init() {
        setContentView(R.layout.activity_start)

        txtSetFreedomDate!!.setOnClickListener { v -> onFreedomDateClick() }
        txtSetFreedomTime!!.setOnClickListener { v -> onFreedomDateTime() }
        btnContinue!!.setOnClickListener { v ->
            if (isDateSet && freedomTime > System.currentTimeMillis()) {
                PrefHelper.setFreedomTime(this@StartActivity, freedomTime)
                PrefHelper.setCountdownStartDate(this@StartActivity, System.currentTimeMillis())
                AlarmHelper.setupFinishingNotification(this@StartActivity)
                navigateCountdownView()
            } else {
                AlertDialog.Builder(this)
                        .setCancelable(true)
                        .setTitle(R.string.date_warning_title)
                        .setMessage(R.string.date_warning_msg)
                        .setPositiveButton(android.R.string.ok) { dialog, which -> dialog.dismiss() }
                        .show()
            }
        }

        txtPickedTime!!.visibility = View.GONE
    }

    private fun onFreedomDateTime() {
        val now = Calendar.getInstance()
        TimePickerDialog(this, this,
                now.get(Calendar.HOUR_OF_DAY),
                now.get(Calendar.MINUTE), true).show()
    }

    private fun onFreedomDateClick() {
        val now = Calendar.getInstance()
        val dialog = DatePickerDialog(this, this,
                now.get(Calendar.YEAR),
                now.get(Calendar.MONTH),
                now.get(Calendar.DAY_OF_MONTH))
        dialog.datePicker.minDate = now.timeInMillis

        now.add(Calendar.YEAR, PrefHelper.MAX_YEARS_AHEAD)
        dialog.datePicker.maxDate = now.timeInMillis

        dialog.show()
    }

    private fun updateFreedomDateView() {
        if (freedomTime != PrefHelper.FREEDOM_TIME_DEFAULT.toLong()) {
            txtPickedTime!!.text = DateFormatHelper.formatForStart(freedomTime, DateFormat.is24HourFormat(this))
            txtPickedTime!!.visibility = View.VISIBLE
        }
    }

    private fun navigateCountdownView() {
        startActivity(Intent(this, CountdownActivity::class.java))
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
    }
}
